tests/wasm/0setup_test.go

```diff
@@ -0,0 +1,227 @@
+package wasm
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/chromedp/cdproto/cdp"
+	"github.com/chromedp/chromedp"
+)
+
+var dockerHeadless = flag.Bool("docker-headless", false, "Launch headless Chrome via docker instead looking for it on local filesystem")
+var noCleanup = flag.Bool("no-cleanup", false, "Skip cleanup operations (useful for manual viewing and debugging)")
+
+func TestMain(m *testing.M) {
+	flag.Parse()
+
+	setup()
+	ret := m.Run()
+	cleanup()
+	os.Exit(ret)
+}
+
+func setup() {
+
+	// copy wasm_exec.js from targets to this dir
+	b, err := ioutil.ReadFile("../../targets/wasm_exec.js")
+	must(err)
+	must(ioutil.WriteFile("wasm_exec.js", b, 0644))
+
+	if *dockerHeadless {
+
+		// build server for arch that works in docker
+		args := []string{"go", "build", "-o", "server", "0setup_server.go"}
+		cmd := exec.Command(args[0], args[1:]...)
+		cmd.Env = append(os.Environ(), "GOARCH=amd64", "GOOS=linux")
+		b, err := cmd.CombinedOutput()
+		log.Printf("Command: %s; err=%v; full output:\n%s", strings.Join(args, " "), err, b)
+		must(err)
+
+		absDir, err := filepath.Abs(".")
+		must(err)
+		// launch headless chrome and server via docker if requested
+		must(run("docker build -t tinygo_wasm_test -f 0dockerfile ."))
+		run("docker stop tinygo_wasm_test") // ignore error
+
+		must(runargs("docker", "run",
+			"-d", "--rm",
+			"--name=tinygo_wasm_test",
+			"--mount", "type=bind,target=/wasm,source="+absDir,
+			"-p", "9222:9222", "-p", "8826:8826",
+			"tinygo_wasm_test:latest"))
+
+	} else {
+
+		killServer()
+
+		// build server for local arch
+		must(run("go build -o server 0setup_server.go"))
+
+		// or otherwise just launch server
+		serverCmd = exec.Command("./server")
+		serverCmd.Stdout = os.Stdout
+		serverCmd.Stderr = os.Stderr
+		must(serverCmd.Start())
+
+		must(ioutil.WriteFile("server.pid", []byte(fmt.Sprintf("%d\n", serverCmd.Process.Pid)), 0644))
+
+	}
+
+}
+
+var serverCmd *exec.Cmd
+
+func cleanup() {
+	// log.Println("performing cleanup")
+
+	if *noCleanup {
+		return
+	}
+	if *dockerHeadless {
+		run("docker stop tinygo_wasm_test") // ignore error
+	} else {
+		killServer()
+	}
+	os.Remove("wasm_exec.js")
+	os.Remove("server")
```

Comment:
Would it be possible to use a temporary directory instead? That avoids the need for remembering to clean up all the files, and doesn't leave any files in the testing directory after the test aborts unexpectedly.

The pattern I usually use is:

```go
dir, err := ioutil.TempDir("", "tinygo-wasm-test")
// handle err
defer os.RemoveAll(dir)

// Run tests, storing temporary files in dir.
```

This pattern cleans up all files even in the case of a panic, but still doesn't leave any around on very unexpected exits (e.g. power loss) in the source directory. The OS will still clean up the temporary files, probably on the next reboot.